src/cloud/linode.rs

```diff
@@ -0,0 +1,168 @@
+use crate::ops::{ExitNode, ExitNodeStatus, EXIT_NODE_PROVISIONER_LABEL};
+use async_trait::async_trait;
+use base64::Engine;
+use color_eyre::eyre::{anyhow, Error};
+use k8s_openapi::api::core::v1::Secret;
+use linode_rs::LinodeApi;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use tracing::{debug, info, warn};
+
+const TOKEN_KEY: &str = "LINODE_TOKEN";
+const INSTANCE_TYPE: &str = "g6-nanode-1";
+const IMAGE_ID: &str = "linode/ubuntu22.04";
+
+fn default_size() -> String {
+    String::from(INSTANCE_TYPE)
+}
+#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
+pub struct LinodeProvisioner {
+    /// Name of the secret containing the Linode API token
+    pub auth: String,
+    /// Region ID of the Linode datacenter to provision the exit node in
+    pub region: String,
+    /// Size for the Linode instance
+    #[serde(default = "default_size")]
+    pub size: String,
+}
+
+impl LinodeProvisioner {
+    // gets token from Secret
+    pub async fn get_token(&self, secret: &Secret) -> color_eyre::Result<String> {
+        let data = secret
+            .data
+            .clone()
+            .ok_or_else(|| Error::msg("No data found in secret"))?;
+        let token = data
+            .get(TOKEN_KEY)
+            .ok_or_else(|| Error::msg("No token found in secret"))?;
+
+        let token = String::from_utf8(token.clone().0)?;
+        Ok(token)
+    }
+}
+
+#[async_trait]
+impl Provisioner for LinodeProvisioner {
+    async fn create_exit_node(
+        &self,
+        auth: Secret,
+        exit_node: ExitNode,
+    ) -> color_eyre::Result<ExitNodeStatus> {
+        let password = generate_password(32);
+
+        let _secret = exit_node.generate_secret(password.clone()).await?;
+
+        let config = generate_cloud_init_config(&password, exit_node.spec.port);
+
+        // Okay, so apparently Linode uses base64 for user_data, so let's
+        // base64 encode the config
+
+        let user_data = base64::engine::general_purpose::STANDARD.encode(config);
+
+        let api = LinodeApi::new(self.get_token(&auth).await?);
+
+        let provisioner = exit_node
+            .metadata
+            .annotations
+            .as_ref()
+            .and_then(|annotations| annotations.get(EXIT_NODE_PROVISIONER_LABEL))
+            .unwrap();
+
+        let name = format!(
+            "{}-{}",
+            provisioner,
+            exit_node.metadata.name.as_ref().unwrap()
+        );
+
+        let mut instance = api
+            .create_instance(&self.region, &self.size)
+            .root_pass(&password)
+            .label(&name)
+            .user_data(&user_data)
+            .tags(vec![format!("chisel-operator-provisioner:{}", provisioner)])
+            .image(IMAGE_ID)
+            .booted(true)
+            .run_async()
+            .await?;
+
+        info!("Created instance: {:?}", instance);
+
+        let instance_ip = loop {
```

Comment:
while these loops are okay for now, this isn't a reliable solution that we want to keep for long

we should ideally have an "ip acquiring" state on the ExitNode, and have a single loop in the operator that polls the relevant provider for matched ExitNodes

that way, if the operator crashes while this wait is happening, we can resume from where we left off, with the node we already created with the provider, otherwise there's the chance that we may leave a random dangling node